src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h

```diff
@@ -28,20 +30,102 @@ namespace crimson::os::seastore::omap_manager {
+    ChildNode<OMapInnerNode, OMapInnerNode, std::string> {
   using OMapInnerNodeRef = TCachedExtentRef<OMapInnerNode>;
-  using internal_iterator_t = const_iterator;
+  using internal_const_iterator_t = const_iterator;
+  using internal_iterator_t = iterator;
+  using parent_node_t = ParentNode<OMapInnerNode, std::string>;
+  using base_child_t = BaseChildNode<OMapInnerNode, std::string>;
+  using child_node_t = ChildNode<OMapInnerNode, OMapInnerNode, std::string>;
+  static constexpr uint32_t CHILD_VEC_UNIT = 128;
 
   explicit OMapInnerNode(ceph::bufferptr &&ptr)
-    : OMapNode(std::move(ptr)) {
-    this->set_layout_buf(this->get_bptr().c_str());
+    : OMapNode(std::move(ptr)),
+      StringKVInnerNodeLayout(get_bptr().c_str()),
+      parent_node_t(0) {
+    this->parent_node_t::sync_children_capacity();
   }
   // Must be identical with OMapInnerNode(ptr) after on_fully_loaded()
   explicit OMapInnerNode(extent_len_t length)
-    : OMapNode(length) {}
+    : OMapNode(length),
+      StringKVInnerNodeLayout(nullptr),
+      parent_node_t(0) {}
   OMapInnerNode(const OMapInnerNode &rhs)
-    : OMapNode(rhs) {
-    this->set_layout_buf(this->get_bptr().c_str());
+    : OMapNode(rhs),
+      StringKVInnerNodeLayout(get_bptr().c_str()),
+      parent_node_t(0) {
+    this->parent_node_t::sync_children_capacity();
+  }
+
+#ifndef NDEBUG
+  bool validate_child(
+    const base_child_t &child,
+    internal_const_iterator_t iter)
+  {
+    return iter.contains(dynamic_cast<const OMapNode&>(child).get_begin());
+  }
+#endif
+
+  iterator begin() {
+    return iter_begin();
+  }
+
+  iterator end() {
+    return iter_end();
+  }
+
+  void do_on_rewrite(Transaction &t, LogicalCachedExtent &extent) final {
+    auto &ext = static_cast<OMapInnerNode&>(extent);
+    this->parent_node_t::on_rewrite(t, ext);
+    if (ext.is_btree_root()) {
+      set_root();
+    } else {
+      unset_root();
+    }
+    auto &other = static_cast<OMapInnerNode&>(extent);
+    this->init_meta(other.get_begin(), other.get_end());
+    this->sync_children_capacity();
+  }
+
+  void prepare_commit() final {
+    this->parent_node_t::prepare_commit();
+    if (is_rewrite() && !is_btree_root()) {
+      auto &prior = *get_prior_instance()->template cast<OMapInnerNode>();
+      if (prior.base_child_t::has_parent_tracker()) {
+	// unlike fixed-kv nodes, rewriting child nodes of the omap tree
+	// won't affect parent nodes, so we have to manually take prior
+	// instances' parent trackers here.
+	this->child_node_t::take_parent_from_prior();
+      }
+    }
+  }
+
+  void do_on_replace_prior() final {
+    this->parent_node_t::on_replace_prior();
+    if (!this->is_btree_root()) {
+      // XXX: On startup, parent/child nodes may not be linked,
+      //      we need to tolerate this case.
+      auto &prior = *get_prior_instance()->template cast<OMapInnerNode>();
+      this->child_node_t::on_replace_prior(
+	prior.base_child_t::has_parent_tracker());
```

Comment:
Why not just:
```c++
if (prior.base_child_t::has_parent_tracker()) {
  this->child_node_t::on_replace_prior();
}
```
without the need to pass an optional boolean ?

Reply:
I thought on_replace_prior may contain other operations in the future, since its meaning is just "do works necessary when replacing priors". So I chose to let `child_node_t::on_replace_prior()` decide for itself and callers just provide necessary information.

Reply:
Currently, this looks unnecessarily complicated. But this is not a very improtant issue.